backend/assignments/1/src/main/java/org/example/Assignment/ExecuteTransaction.java

```diff
@@ -0,0 +1,197 @@
+            } else {
+                trader.getPortfolio().coinQuantities.put(coin.getSymbol(), quantity);
+            }
+
+            if (trader.getPortfolio() != null) {
+                trader.getPortfolio().totalProfitLoss = trader.getPortfolio().totalProfitLoss - (coin.getPrice() * quantity);
+
+            }
+
+        }
+        return true;
+    }
+
+    private boolean sellingFunction() {
+        long quantity = transaction.get("data").get("quantity").asLong();
+        Trader trader = traderMap.get(transaction.get("data").get("wallet_address").asText());
+        Coin coin = coinMap.get(transaction.get("data").get("coin").asText());
+        if (trader == null || trader.getPortfolio() == null) {
+            return false;
+        }
+        Map<String, Long> coinQuantities = trader.getPortfolio().coinQuantities;
+        if (coinQuantities == null) {
+            return false;
+        }
+
+        if (trader.getPortfolio() == null) return false;
+
+
+        synchronized (trader) {
+            if (coin != null && trader.getPortfolio() != null) {
+                Long availableVolume = coin.getCirculatingSupply();
+                double price = coin.getPrice();
+
+                double transactionNet = price * quantity;
+
+                coin.setCirculatingSupply(availableVolume + quantity);
+                if (trader.getPortfolio().coinQuantities.containsKey(coin.getSymbol())) {
+                    Long available = trader.getPortfolio().coinQuantities.get(coin.getSymbol());
+                    Long updated = available - quantity;
+                    if (updated < 0) {
+                        updated = 0L;
+                    }
+                    trader.getPortfolio().coinQuantities.put(coin.getSymbol(), updated);
+                } else {
+                    trader.getPortfolio().coinQuantities.put(coin.getSymbol(), quantity);
+                }
+            }
+        }
+        return true;
+
+    }
+
+
+    private boolean updatePrice() {
+        double newPrice = transaction.get("data").get("price").asDouble();
+        Coin coin = coinMap.get(transaction.get("data").get("coin").asText());
+        if(coin==null) return false;
+        synchronized (coin) {
+
+
+            coin.setPrice(newPrice);
+        }
+
+        return true;
+    }
+
+    private boolean updateVolume() {
+        long newVolume = transaction.get("data").get("volume").asLong();
+        Coin coin = coinMap.get(transaction.get("data").get("coin").asText());
+        if(coin==null) return false;
+        synchronized (coin) {
+
+            coin.setVolume(newVolume);
+        }
+
+        return true;
+
+    }
+
+    @Override
+    public void run() {
+
+        boolean flag = false;
+        int count = 200;
+        while (!flag && count > 0) {
+            if (transactionType.equals("BUY")) {
+
+                flag = buyingFunction();
+                if (flag) {
+                    Log.customLogger("Bought!!", "INFO");
```

Comment:
Making a logger sub-module. Exposing generic methods for logger rather than calling the instances is cleaner and re-usable among projects.